metadata/src/main/java/org/apache/kafka/controller/PartitionChangeBuilder.java

```diff
@@ -328,13 +353,37 @@ public Optional<ApiMessageAndVersion> build() {
 
         completeReassignmentIfNeeded();
 
+        if (PartitionChangeBuilder.this.eligibleLeaderReplicasEnabled) {
+            populateTargetElr();
+        }
+
         tryElection(record);
 
         triggerLeaderEpochBumpIfNeeded(record);
 
-        if (record.isr() == null && !targetIsr.isEmpty() && !targetIsr.equals(Replicas.toList(partition.isr))) {
-            // Set the new ISR if it is different from the current ISR and unclean leader election didn't already set it.
-            record.setIsr(targetIsr);
+        // During the leader election, it can set the record isr if an unclean leader election happens.
```

Comment:
Could we separate the ELR election code into a separate method? Then we can have one branch point where we decided if we need to perform ELR election vs the old code. This would also let us keep the old code unmodified which has some merit.

Reply:
This PR mostly just manages the ELR fields, no election behavior changes. I will try to make the leader election in separate paths with/without ELR in the following leader election PR.

Reply:
As discussed, split the logic to two methods, calling one of them before and after the election.